File: data_structures/week4_hashing/1_phone_book/phone_book.py
from typing import Iterable, List


class Query:
    def __init__(self, query: List[str]) -> None:
        self.type = query[0]
        self.number = int(query[1])

        if self.type == "add":
            self.name = query[2]


def read_queries() -> List[Query]:
    n = int(input())

    return [Query(input().split()) for _ in range(n)]


def process_inefficient(queries: Iterable[Query]) -> List[str]:  # noqa: C901
    contacts: List[Query] = []
    result = []

    for cur_query in queries:
        if cur_query.type == "add":
            for contact in contacts:
                if contact.number == cur_query.number:
                    contact.name = cur_query.name
                    break
            else:
                contacts.append(cur_query)

        elif cur_query.type == "del":
            for j in range(len(contacts)):
                if contacts[j].number == cur_query.number:
                    contacts.pop(j)
                    break

        else:
            response = "not found"
            for contact in contacts:
                if contact.number == cur_query.number:
                    response = contact.name
                    break
            result.append(response)

    return result


def process(queries: List[Query]) -> List[str]:  # noqa: C901
    contact_map = {}
    result = []

    for cur_query in queries:
        if cur_query.type == "add":
            contact_map[cur_query.number] = cur_query.name
        elif cur_query.type == "del":
            contact_map.pop(cur_query.number, None)
        else:
            result.append(contact_map.get(cur_query.number, "not found"))

    return result


def write_responses(result: List[str]) -> None:
    print("\n".join(result))


if __name__ == "__main__":
    write_responses(process(read_queries()))

File: data_structures/week4_hashing/1_phone_book/phone_book.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using std::cin;
using std::string;
using std::vector;

namespace phone_book {
struct Query {
  string type, name;
  int number;
};

vector<Query> read_queries() {
  int n;
  cin >> n;
  vector<Query> queries(n);
  for (int i = 0; i < n; ++i) {
    cin >> queries[i].type;
    if (queries[i].type == "add")
      cin >> queries[i].number >> queries[i].name;
    else
      cin >> queries[i].number;
  }
  return queries;
}

void write_responses(const vector<string>& result) {
  for (size_t i = 0; i < result.size(); ++i)
    std::cout << result[i] << std::endl;
}

vector<string> process_inefficient(const vector<Query>& queries) {
  vector<Query> contacts;
  vector<string> result;

  for (size_t i = 0; i < queries.size(); ++i)
    if (queries[i].type == "add") {
      bool found = false;

      for (size_t j = 0; j < contacts.size(); ++j)
        if (contacts[j].number == queries[i].number) {
          contacts[j].name = queries[i].name;
          found = true;
          break;
        }

      if (!found)
        contacts.push_back(queries[i]);

    } else if (queries[i].type == "del") {
      for (size_t j = 0; j < contacts.size(); ++j)
        if (contacts[j].number == queries[i].number) {
          contacts.erase(contacts.begin() + j);
          break;
        }

    } else {
      string response = "not found";
      for (size_t j = 0; j < contacts.size(); ++j)
        if (contacts[j].number == queries[i].number) {
          response = contacts[j].name;
          break;
        }
      result.push_back(response);
    }

  return result;
}

vector<string> process(const vector<Query>& queries) {
  std::unordered_map<int, string> contact_map;
  vector<string> result;

  for (const Query& cur_query : queries) {
    if (cur_query.type == "add") {
      contact_map[cur_query.number] = cur_query.name;
    } else if (cur_query.type == "del") {
      contact_map.erase(cur_query.number);
    } else {
      auto it = contact_map.find(cur_query.number);
      if (it != contact_map.end()) {
        result.push_back(it->second);
      } else {
        result.push_back("not found");
      }
    }
  }

  return result;
}
}  // namespace phone_book

int main() {
  phone_book::write_responses(phone_book::process(phone_book::read_queries()));
  return 0;
}
